libsolidity/analysis/TypeChecker.cpp

```diff
@@ -338,35 +340,59 @@ bool TypeChecker::visit(FunctionDefinition const& _function)
 		if (_function.isOrdinary() && !_function.isPartOfExternalInterface())
 			m_errorReporter.typeError(5587_error, _function.location(), "Internal functions cannot be payable.");
 	}
+
+	vector<VariableDeclaration const*> internalParametersInConstructor;
+
 	auto checkArgumentAndReturnParameter = [&](VariableDeclaration const& var) {
 		if (type(var)->containsNestedMapping())
 			if (var.referenceLocation() == VariableDeclaration::Location::Storage)
 				solAssert(
-					_function.libraryFunction() || !_function.isPublic(),
+					_function.libraryFunction() || _function.isConstructor() || !_function.isPublic(),
 					"Mapping types for parameters or return variables "
 					"can only be used in internal or library functions."
 				);
-		if (_function.isPublic())
+		bool functionIsExternallyVisible =
+			(!_function.isConstructor() && _function.isPublic()) ||
+			(_function.isConstructor() && !m_currentContract->abstract());
+		if (
+			_function.isConstructor() &&
+			var.referenceLocation() == VariableDeclaration::Location::Storage
+		)
+			// TODO what about other things like internal function pointers?
```

Comment:
Yea, that's tricky. At a first glance I would say they're rather similar to "normal" state variables

Reply:
The case is covered here: https://github.com/ethereum/solidity/pull/9180/commits/d608952f1116d988b1d1ed07dfc3e4edc4103ad4#diff-af6ececd486e0ecba425b992a1755ce8R2
I will remove the TODO.

Reply:
But actually it is weird - the two errors should be unified.